src/puppetlabs/puppetdb/threadpool.clj

```diff
@@ -0,0 +1,130 @@
+  (.shutdown threadpool)
+
+  ;; This will block, waiting for the threadpool to shutdown
+  (when-not (.awaitTermination threadpool shutdown-timeout java.util.concurrent.TimeUnit/MILLISECONDS)
+
+    (log/warn
+     (i18n/trs "Threadpool not stopped after {0} milliseconds, forcibly shutting it down" shutdown-timeout))
+
+    ;; This will force the shutdown of the threadpool and will
+    ;; not allow current threads to finish
+    (.shutdownNow threadpool)
+    (log/warn
+     (i18n/trs "Threadpool forcibly shutdown"))))
+
+(defrecord GatedThreadpool [^Semaphore semaphore ^ExecutorService threadpool shutdown-timeout]
+  java.io.Closeable
+  (close [this]
+    (shutdown this)))
+
+(defn create-threadpool
+  "Creates an unbounded threadpool with the intent that access to the
+  threadpool is bounded by the semaphore. Implicitly the threadpool is
+  bounded by `size`, but since the semaphore is handling that aspect,
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size name-pattern shutdown-timeout-in-ms]
+  (let [threadpool (ThreadPoolExecutor. 1
+                                        Integer/MAX_VALUE
+                                        1
+                                        TimeUnit/MINUTES
+                                        (SynchronousQueue.))]
+    (->> threadpool
+         .getThreadFactory
+         (thread-factory name-pattern)
+         (.setThreadFactory threadpool))
+
+    (map->GatedThreadpool
+     {:semaphore (Semaphore. size)
+      :threadpool threadpool
+      :shutdown-timeout shutdown-timeout-in-ms})))
+
+(defn call-on-threadpool
+  "Executes `f` on a gated threadpool. Calls `on-complete` with the
+  results of inoking `f`"
+  [{:keys [^Semaphore semaphore ^ExecutorService threadpool] :as gated-threadpool}
+   f]
+  (.acquire semaphore)
+  (try
+    (.execute threadpool (fn []
+                           (try
+                             (f)
+                             (finally
+                               (.release semaphore)))))
+    ;; RejectedExecutionExceptions only occur when attempting ot
+    ;; excecute new work on a threadpool that is in the process of
+    ;; shutting down or already has been shut down
+    (catch RejectedExecutionException e
+      (.release semaphore)
+      (throw e))))
+
+(defn exec-from-channel
+  "Executes `on-input` for each input found on `in-chan`, with a
+  threadpool of size `parallelism`. The results of the `on-input`
+  invocations will be put to the `out-chan` in the order it has
+  completed. Throws when the threadpool is shutting down and there are
+  still messages in the channel"
+  [gated-threadpool in-chan on-input]
+  (loop [cmd (async/<!! in-chan)]
+    (when cmd
+      (try
+        (call-on-threadpool
+         gated-threadpool
+         (fn []
+           (on-input cmd)))
+        (catch RejectedExecutionException e
+          (throw+ {:kind ::rejected
```

Comment:
Catching it on line 101 is so that we can release the semaphore. An argument could be made for NOT releasing the sempahore here. Happy to talk through that, but it's an edge case that we shouldn't hit. We catch it here so that we can provide context, like the specific command that failed. This is something that shouldn't happen if we're using this correctly, but I figured if it did, you'd want to know which command didn't get processed.